pandas/tests/tseries/offsets/test_behavior.py

```diff
@@ -0,0 +1,334 @@
+                                    MonthEnd, MonthBegin,
+                                    BMonthEnd, BMonthBegin,
+                                    QuarterEnd, QuarterBegin,
+                                    BQuarterEnd, BQuarterBegin,
+                                    YearEnd, YearBegin,
+                                    BYearEnd, BYearBegin,
+                                    Week, LastWeekOfMonth, WeekOfMonth,
+                                    SemiMonthBegin, SemiMonthEnd,
+                                    Easter,
+                                    FY5253, FY5253Quarter,
+                                    DateOffset)
+# TODO:
+# BusinessDay, BusinessHour, CustomBusinessDay, CustomBusinessHour,
+# CustomBusinessMonthEnd, CustomBusinessMonthBegin
+
+
+tick_classes = [Hour, Minute, Second, Milli, Micro, Nano]
+yqm_classes = [MonthBegin, MonthEnd, BMonthBegin, BMonthEnd,
+               QuarterBegin, QuarterEnd, BQuarterBegin, BQuarterEnd,
+               YearBegin, YearEnd, BYearBegin, BYearEnd]
+offset_types = [Week, LastWeekOfMonth, WeekOfMonth, SemiMonthEnd,
+                SemiMonthBegin, FY5253Quarter, FY5253,
+                Easter, DateOffset] + tick_classes + yqm_classes
+
+# ----------------------------------------------------------------
+# Helpers for generating random data
+
+dt_max = pd.Timestamp.max.replace(nanosecond=0).to_pydatetime()
+td_max = timedelta(106751, 85636, 854775)
+td_min = -td_max - timedelta(microseconds=1)
+
+n_strategy = st.integers(min_value=-999, max_value=999)
+# TODO: Choose these bounds systematically.  (-999, 999) is arbitrarily chosen
+# to get rid of OverflowErrors in development
+month_strategy = st.integers(min_value=1, max_value=12)
+weekday_strategy = st.integers(min_value=0, max_value=6)
+
+
+def gen_dst_crossing():
+    # Generate either a pair of Timestamps or a date_range that is known
+    # to cross a DST transition
+    raise NotImplementedError
+
+
+def gen_date_range_freq():
+    # return a freq str or offset object suitable for passing as
+    # `freq` kwarg to date_range
+    return st.sampled_from(['Y', 'Q', 'M', 'D', 'H',
+                            'T', 's', 'ms', 'us', 'ns'])
+    # TODO: Add the rest; business, multiples, ...
+
+
+@st.composite
+def gen_random_date_range(draw):
+    # TODO: Choose the min/max values more systematically
+    start = st.datetimes(min_value=pd.Timestamp(1900, 1, 1).to_pydatetime(),
+                         max_value=pd.Timestamp(2100, 1, 1).to_pydatetime())
+    periods = st.integers(min_value=10, max_value=100)
+    freq = gen_date_range_freq()
+    tz = gen_random_tz()
+
+    dti = pd.date_range(start=draw(start), tz=draw(tz),
+                        freq=draw(freq), periods=draw(periods))
+    return dti
+
+
+def gen_random_tz():
+    # Allows None
+    return st.one_of(st.none(), hepytz.timezones())
+    # TODO: Weighting between naive and timezones?
+    # TODO: Get datetuil timezones?
```

Comment:
I have been briefly in contact with @DRMacIver about adding a `hypothesis.extras.dateutil` extra to do just this. It *may* require some additional public interface stuff on the dateutil side.

That said, it seems that all timezones provided by `pytz` map identically to `dateutil` zones, for all datetimes less than 2038-01-01. To prove it, here's a hypothesis test I ran:

```python
from hypothesis import given, assume
from hypothesis import strategies as st
from hypothesis.extra import pytz as hepytz

from dateutil import tz
from datetime import datetime

@given(dt=st.datetimes(), tzi=hepytz.timezones())
def test_dateutil_compat(dt, tzi):
    tzi_du = tz.gettz(str(tzi))
    dt_pytz = tzi.localize(dt)
    dt_du = dt.replace(tzinfo=tzi_du)

    assume(dt < datetime(2038, 1, 1))
    assert dt_pytz == dt_du
```

So you should be able to get `dateutil` zones from `tz.gettz(str(pytz_zone))`

Reply:
Also, does this need to be a function? Why not just assign `timezone_strategy = st.one_of([None, hepytz.timezones()])`?

Reply:
> Also, does this need to be a function?

Probably not.  I'm still getting the hang of `@composite` and, `data`, and `draw`.

Reply:
> So you should be able to get dateutil zones from tz.gettz(str(pytz_zone))

That looks great, thanks.  I'll take a look at dateutil-590 to see if I can be helpful.